### PR TITLE
OTLP: Support receiving logs in Loki using OpenTelemetry OTLP

### LINKED ISSUE #5346: Support receiving logs in Loki using OpenTelemetry OTLP
**Is your feature request related to a problem? Please describe.**
I am running Grafana Loki inside a Kubernetes cluster but I have some applications running _outside_ the cluster and I want to get logging data from those applications into Loki without relying on custom APIs or file-based logging.

**Describe the solution you'd like**
[OpenTelemetry describes a number of approaches](https://opentelemetry.io/docs/reference/specification/logs/overview/) including using the OpenTelemetry Collector. The OpenTelemetry Collector [supports various types of exporters](https://opentelemetry.io/docs/collector/configuration/#exporters) and the OTLP exporter supports logs, metrics, and traces. Tempo supports receiving trace data via OTLP and it would be great if Loki also had support for receiving log data via OTLP. This way, people could run the OpenTelemetry Collector next to their applications and send logs into Loki in a standard way using the [OpenTelemetry New First-Party Application Logs recommendations](https://opentelemetry.io/docs/reference/specification/logs/overview/#new-first-party-application-logs).

Currently, unless I am misunderstanding the Loki documentation, it seems the only API into Loki is custom:
- [POST /loki/api/v1/push](https://grafana.com/docs/loki/latest/api/#post-lokiapiv1push)

Details on the OTLP specification:
- OTLP/gRPC: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlpgrpc
- OTLP/HTTP: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp

**Describe alternatives you've considered**
There are a number of [Loki Clients](https://grafana.com/docs/loki/latest/clients/) that one can use to get logs into Loki but they all seem to involve using the custom Loki push API or reading from log files. Supporting the OpenTelemetry Collector would allow following the [OpenTelemetry New First-Party Application Logs recommendations](https://opentelemetry.io/docs/reference/specification/logs/overview/#new-first-party-application-logs)

**Additional context**

### PR DESCRIPTION
**Pull request by jpkrohling**
**What this PR does / why we need it**:
This change brings support to ingesting OTLP as part of the main
ingester. This uses the standard OpenTelemetry Collector pdata
Logs gRPC server as the interface and the translator package
from the contrib repository. Once the data is translated from OTLP
to a Loki's PushRequest, we call use the Ingester's Push function.

**Which issue(s) this PR fixes**:
Fixes #5346

**Special notes for your reviewer**:
Most of this change is about bringing new dependencies. The code itself is quite simple, just making bridges. I tested this with an OpenTelemetry Collector Contrib with the following configuration:

```yaml
receivers:
  journald:
    directory: /var/log/journal/a04e3a44cdd740f88d6a7ae3bb8c70cf

exporters:
  logging:
    loglevel: debug
  otlp:
    endpoint: http://localhost:9095
    tls:
      insecure: true

processors:
  attributes:
    actions:
    - action: insert
      key: loki.attribute.labels
      value: http_status_code

    # the following attributes are added manually here in the example, but would
    # probably be added by other processors or straight from the source
    - action: insert
      key: http_status_code
      value: 500

  resource:
    attributes:
    - action: insert
      key: loki.resource.labels
      value: host_name, pod_name

    # the following attributes are added manually here in the example, but would
    # probably be added by other processors or straight from the source
    - action: insert
      key: host_name
      value: guarana
    - action: insert
      key: pod_name
      value: guarana-pod-01

extensions:

service:
  extensions:
  pipelines:
    logs:
      receivers: [journald]
      processors: [resource, attributes]
      exporters: [logging, otlp]
```

**Checklist**
- [x] Reviewed the `CONTRIBUTING.md` guide
- [ ] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
- [x] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by jpkrohling** [comment]
I'm not sure what's the best place to document this. There aren't configuration options, the only things to mention would be:

- state that OTLP is accepted in the main gRPC port
- provide an example configuration file, like the one I posted in the issue description

**Comment by jpkrohling** [comment]
There's something wrong with the CLA check. I signed, it says I'm signed, but the check is still failing:

image

File: pkg/otlp/otlp.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	"go.opentelemetry.io/collector/pdata/plog/plogotlp"
+
+	"github.com/grafana/loki/pkg/logproto"
+	util_log "github.com/grafana/loki/pkg/util/log"
+)
+
+var _ plogotlp.Server = (*Server)(nil)
+
+// Server is an OpenTelemetry Collector Logs gRPC server which converts OTLP
+// into Loki's PushRequest and forwards the requeste to the provided PusherServer.
```
Comment (MichelHollands) [inline pkg/otlp/otlp.go:18]:
```suggestion
// into Loki's PushRequest and forwards the request to the provided PusherServer.
```

File: go.mod [hunk 2]
```diff
@@ -54,7 +54,7 @@ require (
 	github.com/grafana/regexp v0.0.0-20220304100321-149c8afcd6cb
 	github.com/grpc-ecosystem/go-grpc-middleware v1.3.0
 	github.com/grpc-ecosystem/grpc-opentracing v0.0.0-20180507213350-8e809c8a8645
-	github.com/hashicorp/consul/api v1.13.0
+	github.com/hashicorp/consul/api v1.15.2
```
Comment (MichelHollands) [inline go.mod:57]: Are all these version changes needed for oltp? If not I'd prefer it if they were in a separate PR.

**Comment by jpkrohling** [comment]
I'm closing this as we decided not to pursue this at the moment.

**Comment by evantorrie** [comment]
> I'm closing this as we decided not to pursue this at the moment.

This is still something that is high on our list of desired features.

Is there still a desire on Grafana's part to deliver native OTLP ingestion for Loki?

**Comment by jpkrohling** [comment]
> Is there still a desire on Grafana's part to deliver native OTLP ingestion for Loki?

Right now, we are working on a proxy for Grafana Cloud that would be able to ingest all three signals. Native OTLP ingestion for Loki will eventually land here, but some other base work has to be completed before.

**Comment by slim-bean** [comment]
Native OTLP ingestion is definitely on our list, but we are still working out how best to store the additional structured data sent with OTLP.

The `resource` data should map well to labels in Loki, however the `attributes` and other fields like `trace` and `span` may/will be too high cardinality to put in the index.

This PR takes the same approach as the [Loki exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/lokiexporter/README.md) which allows selective mapping of attributes to labels but this only will work for low cardinality attributes.

We recommend using the collector and exporter until we can natively support and store all the data in the OTLP payload directly in Loki, however to do this we need to make some store changes to make a place for structured data.

**Comment by feldentm-SAP** [comment]
@slim-bean are there any plans to extend the Loki data store to handle attributes like traceID/spanID or other correlators efficiently? I mean this would be required before dumping all the OTLP-compliant data into Loki.

**Comment by slim-bean** [comment]
Yes there are plans for this! Most of us are just getting back up to speed after some holiday time off and I know there are a few meetings scheduled to talk about this in the coming weeks.

Optimistically I hope we can have something started in Q1 this year.

**Comment by frzifus** [comment]
Is there some kind of public roadmap?

**Pull request closed without merging**